File: design-patterns/golang/observer/goal_notifier.go
package observer

import "fmt"

// GoalNotifier notifies when a goal is reached
type GoalNotifier struct {
	goalSteps int
}

func NewGoalNotifier(goalSteps int) *GoalNotifier {
	return &GoalNotifier{goalSteps: goalSteps}
}

func (g *GoalNotifier) Update(steps int, calories int, distance float64) {
	if steps >= g.goalSteps {
		fmt.Printf("[GoalNotifier] Congratulations! Goal of %d steps reached!\n", g.goalSteps)
	} else {
		fmt.Printf("[GoalNotifier] %d steps to go to reach your goal of %d steps.\n", g.goalSteps-steps, g.goalSteps)
	}
} 

File: design-patterns/python/observer/goal_notifier.py
from .fitness_data_observer import FitnessDataObserver


class GoalNotifier(FitnessDataObserver):
    """Concrete Observer - notifies when goals are achieved"""
    
    def __init__(self, step_goal: int = 10000, calorie_goal: float = 500.0):
        self._step_goal = step_goal
        self._calorie_goal = calorie_goal
    
    def update(self, fitness_data) -> None:
        if fitness_data.steps >= self._step_goal:
            print(f"GoalNotifier: Congratulations! You've reached your step goal of {self._step_goal} steps!")
        
        if fitness_data.calories_burned >= self._calorie_goal:
            print(f"GoalNotifier: Great job! You've burned {self._calorie_goal} calories!")